Timeout can be specified for roscpp service calls

dkargin opened the pull request:
Added function ros::ServiceClient::setTimeout to limit the time roscpp service client waits for response. By defaut it is set to -1.0 seconds,  and the service call blocks until it gets a response or connection is dropped.

Straightforward solution for #152

dkargin committed:
Timeout can be specified for roscpp service calls

Added function ros::ServiceClient::setTimeout to limit the time
roscpp service client waits for response. By defaut it is set
to -1 and service call blocks it gets a response or connection
is dropped.

evenator commented:
I'm not an expert on C++ linkers, but I think this would cause an ABI change, which would make it ineligible for released ROS distros, especially indigo. I hope someone with more knowledge on C++ ABIs than me could come up with a way to implement this that doesn't change the ABI, because it would be a useful feature. Barring that, I hope this PR will be accepted for KTurtle.

dkargin commented:
It 'should'  be fine for ABI. I've added a new variable into Impl class. And new class method should not break C++ ABI. There could be a problem with internal ABI, because I've added some new arguments. Does anybody uses that internal stuff directly? I mean ServiceServerLink::call method.

But I've found some other pull requests, i.e for solving 'reconnect' behaviour for service client, like #574 . The final solution, that supports both reconnect and timeout functionality, should be a bit more complex.

dirk-thomas commented on clients/roscpp/src/libros/service_client.cpp:
```diff
@@ -172,6 +172,14 @@ bool ServiceClient::isPersistent() const
+	if(impl_)
+	{
+		impl_->timeout = timeout;
+	}
```
Please follow the style in the file ans use two spaces for indentation.

Same below.

dirk-thomas commented on clients/roscpp/src/libros/service_server_link.cpp:
```diff
@@ -357,15 +358,30 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
   }
 
+  if(interrupted)
+	  this->clearCalls();
```
This function call seems to require the same lock as used above: `info->finished_mutex_`

dkargin replied:
clearCalls is mutex-protected by itself. Protecting it by info->finished_mutex_ will cause a deadlock at ServiceServerLink::cancelCall

dkargin committed:
Fixing for a merge

- fixed style
- call interruption is properly wrapped by a mutex

dkargin committed:
Using wait_untill instead wait_for

Service timeout check is more strict right now. It checks wall time instead of trying to iterate with limited period

dkargin committed:
Fixed deadlock in service call

efernandez commented on clients/roscpp/include/ros/service_server_link.h:
```diff
@@ -105,7 +105,7 @@ class ROSCPP_DECL ServiceServerLink : public boost::enable_shared_from_this<Serv
    * it has finished.
    */
-  bool call(const SerializedMessage& req, SerializedMessage& resp);
+  bool call(const SerializedMessage& req, SerializedMessage& resp, double timeout = -1);
```
This will break ABI compatibility (see http://www.ros.org/reps/rep-0009.html#what-s-not-allowed), but it's very easy to keep ABI compatibility in this case, simply create a new method with the new argument, so you have something like this:

``` c++
bool call(const SerializedMessage& req, SerializedMessage& resp);
bool call(const SerializedMessage& req, SerializedMessage& resp, double timeout);
```

I'm hesitant of providing the new implementation for the 2-argument version here because IIRC if the compiler makes it inline it'd also break ABI compatibility. Therefore, I'd add the implementation on the .cpp.

efernandez commented on clients/roscpp/src/libros/service_server_link.cpp:
```diff
@@ -321,7 +322,7 @@ void ServiceServerLink::processNextCall()
 }
 
-bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& resp)
+bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& resp, double timeout)
```
Following my previous comment above, here you'll have to do this:

``` c++
bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& resp)
{
  return call(req, resp, -1);
}

bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& resp, double timeout)
```

dkargin replied:
Thanks for notice. Fixed

efernandez reviewed (commented):
Consider my coments if you want to keep ABI compatibility.

efernandez commented:
@dkargin I've made my ABI comments just a review "Comment" rather than a request for changes because you're right that those headers are sort of internal, but also note they're exported here https://github.com/sybotech/ros_comm/blob/e5e8d8ac3ffbc1b36ebd9f8bb3a1bf6067aed3a3/clients/roscpp/CMakeLists.txt#L133

dkargin committed:
Fixed ABI problems

dkargin committed:
Trying to ease merging process

efernandez commented:
@dkargin Regarding the merging conflicts, I think all you have to do is:

1. Remove the "Trying to ease merging process" commit
2. From your branch, run `git rebase indigo-devel`
3. Then `git push -f sybotech ServiceCallTimeout` assuming `sybotech` is how you call your fork locally

Maybe you should target `lunar-devel` too, BTW.

efernandez reviewed (approved):
Thanks for applying the changes after my comments. LGTM.

efernandez commented:
Although this change should be ABI compatible now, I'd rather target `lunar-devel` instead of `indigo-devel`.

@dirk-thomas Thoughts?

It's also unclear to me what was the end of the story for #574

mikepurvis commented:
Rebased for `lunar-devel` in #1165. Let's take the discussion there.

peci1 commented:
But we'd also like to have it in Indigo =)

efernandez commented:
IMHO this should go to the latest devel branch (`lunar-devel`) @peci1 and once it gets merged there we can discuss about cherry-picking it into previous versions. I don't have anything against that, but remember that indigo is 3 versions old.

prateek470 commented:
@dkargin Do we have SetTimout in any ROS version?
I am stuck on this one for a long time now. Please merge these changes.

dkargin commented:
@prateek470 No, timeouts are not in the mainline now. You can fork  ros_comm, and adapt this changes and rebuild everything from scratch (like in meta-ros), or do not use ROS services at all, if you need sort of reliability and robustness.
